src/tools/tidy/src/features.rs

```diff
@@ -110,21 +107,105 @@ pub fn check(path: &Path, bad: &mut bool) {
                                     since: since.to_owned(),
+                                    has_gate_test: false,
                                 });
         }
     });
 
+    super::walk(&path.join("test/compile-fail"),
+                &mut |path| super::filter_dirs(path),
+                &mut |file| {
+        let filename = file.file_name().unwrap().to_string_lossy();
+        if !filename.ends_with(".rs") || filename == "features.rs" ||
+           filename == "diagnostic_list.rs" {
+            return;
+        }
+
+        let filen_underscore = filename.replace("-","_").replace(".rs","");
+        test_filen_gate(&filen_underscore, &mut features);
+
+        contents.truncate(0);
+        t!(t!(File::open(&file), &file).read_to_string(&mut contents));
+
+        for (i, line) in contents.lines().enumerate() {
+            let mut err = |msg: &str| {
+                println!("{}:{}: {}", file.display(), i + 1, msg);
+                *bad = true;
+            };
+
+            let gate_test_str = "gate-test-";
+
+            if !line.contains(gate_test_str) {
+                continue;
+            }
+
+            let feature_name = match line.find(gate_test_str) {
+                Some(i) => {
+                    &line[i+gate_test_str.len()..line[i+1..].find(' ').unwrap_or(line.len())]
+                },
+                None => continue,
+            };
+            let found_feature = features.get_mut(feature_name)
+                                        .map(|v| { v.has_gate_test = true; () })
+                                        .is_some();
+
+            let found_lib_feature = features.get_mut(feature_name)
+                                            .map(|v| { v.has_gate_test = true; () })
+                                            .is_some();
+
+            if !(found_feature || found_lib_feature) {
+                err(&format!("gate-test test found referencing a nonexistent feature '{}'",
+                             feature_name));
+            }
+        }
+    });
+
+    // FIXME get this whitelist empty.
+    let whitelist = vec![
+        "abi_ptx", "simd", "safe_suggestion", "macro_reexport",
+        "more_struct_aliases", "static_recursion", "reflect",
+        "quote", "cfg_target_has_atomic", "custom_attribute",
+        "default_type_parameter_fallback", "pushpop_unsafe",
+        "use_extern_macros", "staged_api", "const_indexing",
+        "unboxed_closures", "stmt_expr_attributes",
+        "cfg_target_thread_local", "unwind_attributes",
+        "inclusive_range_syntax"
+    ];
+
+    // Only check the number of lang features.
+    // Obligatory testing for library features is dumb.
+    let gate_untested = features.iter()
+                                .filter(|&(_, f)| f.level == Status::Unstable)
+                                .filter(|&(_, f)| !f.has_gate_test)
+                                .filter(|&(n, _)| !whitelist.contains(&n.as_str()))
```

Comment:
I'm happy to land it as is, but I have a mild concern that people will add tests for something in the whitelist, but forget to remove it from the whitelist. Either we have to be vigilant, or we could tweak this code such that if it is in the whitelist, we error out if it is found to be tested.